Fix optional field handling in MOL2Parser and MOL2

Mrqeoqqt opened the pull request:
Fixes #3385 

Changes made in this Pull Request:
- Modified MOL2Parser: when optional fields are not provided: 
  + `1` assigned to `subst_id`
  + If all atoms have `subst_name`, resnames attribute will be set;
     if no atoms have `subst_name`, resnames attribute will not be set;
     otherwise a ValueError will raise.
  + If "NO_CHARGES" exists in `MOLECULE`  section, and no atoms have `charge` field, charges attribute will not be set;
     If "NO_CHARGES" exists in `MOLECULE`  section while some atoms have `charge` field, OR charge model exists in 
     `MOLECULE`  section while some atoms don't have "charge" field, a ValueError will raise.
- Modified MOL2.py to unpack only compulsory keywords as well.
- Therefore the situation described in #3385 can be properly handled.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

Mrqeoqqt committed:
modified MOL2Parser.py and MOL2.py to unpack only compulsory keywords

Mrqeoqqt committed:
change default values for optional keywords in MOL2Parser.py

Mrqeoqqt committed:
fix PEP in modification

Mrqeoqqt committed:
delete two imports

pep8speaks commented:
Hello @Mrqeoqqt! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`package/MDAnalysis/topology/MOL2Parser.py`](https://github.com/MDAnalysis/mdanalysis/blob/88a3d0116b7c364e50bc4e71c0669e0d22ae2860/package/MDAnalysis/topology/MOL2Parser.py):

> [Line 95:80](https://github.com/MDAnalysis/mdanalysis/blob/88a3d0116b7c364e50bc4e71c0669e0d22ae2860/package/MDAnalysis/topology/MOL2Parser.py#L95): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 98:1](https://github.com/MDAnalysis/mdanalysis/blob/88a3d0116b7c364e50bc4e71c0669e0d22ae2860/package/MDAnalysis/topology/MOL2Parser.py#L98): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 100:80](https://github.com/MDAnalysis/mdanalysis/blob/88a3d0116b7c364e50bc4e71c0669e0d22ae2860/package/MDAnalysis/topology/MOL2Parser.py#L100): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)





##### Comment last updated at 2022-04-08 16:14:03 UTC

Mrqeoqqt committed:
fix the scenario when the number of input fields is greater than 9

Mrqeoqqt committed:
avoid loop when the number of fields is greater or equal to 9

Mrqeoqqt committed:
fix PEP8

Mrqeoqqt committed:
fix PEP8 issur

micaela-matta commented:
Looks good, thanks! Before we merge you would need to:

- update the CHANGELOG
- add a test (you can copy the current MOL2 file into a new one that doesn't have the optional columns, and see that it's being read properly)
- add yourself to AUTHORS since it's your first contribution
- add versionchanged tags to all affected functions

Mrqeoqqt committed:
add a test MOL2 file that doesn't have the optional columns

Mrqeoqqt committed:
add file path in datafiles.py

Mrqeoqqt committed:
Append in changelog about mol2 parser under v2.2.0

Mrqeoqqt committed:
add test function test_optional_columns to test_mol2.py

Mrqeoqqt committed:
delete change in data

Mrqeoqqt committed:
append in AUTHORS

Mrqeoqqt commented:
Hi,

Thank tou for the comments. I am a little confused about what I should do to deal with this one. I guess it would be something related to adding a tag to my most recent push. But I am not sure whether I have understood it right. Could you provide more details? 
> * add versionchanged tags to all affected functions

Mrqeoqqt committed:
modify CHANGELOG

Mrqeoqqt committed:
add newlines in test_mol2.py to fix PEP8 issue

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -150,8 +150,28 @@ def parse(self, **kwargs):
+                                         os.path.basename(self.filename)))
+                aid, name, x, y, z, atom_type = columns[:6]
+                for i in range(6, len(columns)):
+                    opt_values[i-6] = columns[i]
```
This code should be covered by tests.

Mrqeoqqt replied:
Added a testcase with some optional columns.

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -150,8 +150,28 @@ def parse(self, **kwargs):
+                opt_values = ['1', 'UNK', '0.0']
+                opt_fields = ['subst_id', 'subst_name', 'charge']
+                if len(columns) < 6:
+                    raise ValueError("The @<TRIPOS>ATOM block in mol2 file {0}"
```
add test

Mrqeoqqt replied:
Added a testcase with some compulsory columns missing.

orbeckst reviewed (changes requested):
There are some new lines of code that seem important for the new functionality that are not yet covered by tests. Please test all code paths.

For the "versionchanged": look in the doc string of the class/method. Add something like
```reST
.. versionchanged: 2.2.0
   ... summarize new/changed functionality here ...
```

Mrqeoqqt committed:
add more test functions to cover the appended code

Mrqeoqqt committed:
add versionchanged 2.2.0 tags in MOL2Parser and MOL2

Mrqeoqqt committed:
fix PEP8

Mrqeoqqt committed:
fix PEP8

Mrqeoqqt committed:
fix test function test_mol2_wo_required_columns

Mrqeoqqt committed:
fix test_functions

Mrqeoqqt commented:
@micaela-matta @orbeckst Thank you for the comments and explanation.
I tried my best to write the test functions, even though they may not be very comprehensive. I also added versionchanged tags to functions I touched.

orbeckst commented:
@RMeli would you be able to have a quick look at this PR as you raised the original issue? Thanks!

(I’ll have another look tomorrow, just no time available today.)

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -150,8 +152,28 @@ def parse(self, **kwargs):
+                columns = columns[:9]
+                aid, name, x, y, z, atom_type, resid, resname, charge = columns
+            else:
+                opt_values = ['1', 'UNK', '0.0']
+                opt_fields = ['subst_id', 'subst_name', 'charge']
```
I think optional values should not result in attributes, but should be discarded completely if not available. With this solution, you will end up with a `charges` attribute with all `0.0`, for example.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -150,8 +152,28 @@ def parse(self, **kwargs):
+                                     " should contain at least 6 fields to be"
+                                     " unpacked: atom_id atom_name x y z"
+                                     " atom_type [subst_id[subst_name"
+                                     " [charge [status_bit]]]]".format(
+                                         os.path.basename(self.filename)))
```
Please use [f-strings](https://realpython.com/python-f-strings/).

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -150,8 +152,28 @@ def parse(self, **kwargs):
+                aid, name, x, y, z, atom_type = columns[:6]
+                for i in range(6, len(columns)):
+                    opt_values[i-6] = columns[i]
+                for i in range(len(columns), 9):
+                    warnings.warn("Not enough values to unpack."
+                                  " {0} has been given value {1}.".format(
+                                      opt_fields[i-6], opt_values[i-6]))
```
See comment above about discarding attributes instead of providing arbitrary default values. If the MOL2 file is standard-compliant it should not throw warnings.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -150,8 +152,28 @@ def parse(self, **kwargs):
+            else:
+                opt_values = ['1', 'UNK', '0.0']
+                opt_fields = ['subst_id', 'subst_name', 'charge']
+                if len(columns) < 6:
```
This should probably be in its own block:
```python
if len(columns) >= 9:
    pass
elif len(columns) < 6:
    pass
else:
   pass
```

RMeli reviewed (commented):
Some comments after a quick look. I need to re-read carefully the MOL2 standard, which I don't have time to do today. I will give it a deeper look tomorrow afternoon.

Mrqeoqqt committed:
default values of optional fields and fix fstring

Mrqeoqqt commented:
Hi @RMeli ,

Thank you for the comments. I fixed the fstring format. 

But the default value issue is more complicated than I thought. If default values are not provided, interfaces to construct topology in MOL2Parser will not be compatible with others, especially the `resid` and `charge` should be real numbers instead of empty string. 

For now, I can set `resname` to empty string, and leave the charge to `guesser.guess_atom_charge(name)`, but I don't know what to do with `resid`. Besides, when to set charges using `guesser` funstions, it will be a little wired because part of charges are real charges and part of them are guessed. This would cause the problem whether or not pass `guessed=True` when construct the attributes.

Do you have any suggestions?

Mrqeoqqt committed:
Update MOL2Parser.py

orbeckst commented:
Discussion on how to handle missing charges started in #3385 — that's an important thing to decide.

Mrqeoqqt committed:
Update MOL2Parser.py

Mrqeoqqt committed:
don't set charges only when there's 'no_charges' flag or any atom without charge

Mrqeoqqt committed:
parent 3b4baedf060d679e8eb9ca2e2746ddfe77dc9902
author Mingyi Xue <2415309890@qq.com> 1649137066 -0500
committer Mingyi Xue <2415309890@qq.com> 1649148290 -0500

Rebase: net effect: Don't set charge attr when 'no_charges' present

Mrqeoqqt committed:
compare lowercase 'no_charge' in strip string

Mrqeoqqt committed:
compare lowercase 'no_charge' in strip string

Delete affinityprop.c

Delete cutils.c

Delete _augment.cpp

Delete qcprot.c

Delete _cutil.cpp

Delete c_distances_openmp.c

Delete libdcd.c

Delete c_distances.c

Delete libmdaxdr.c

Delete nsgrid.cpp

Delete stochasticproxembed.c

Delete cython_util.c

Mrqeoqqt committed:
Rebase: clean up commits

Mrqeoqqt committed:
Delete affinityprop.c

Mrqeoqqt committed:
Delete cutils.c

Mrqeoqqt committed:
Delete _augment.cpp

Mrqeoqqt committed:
Delete qcprot.c

Mrqeoqqt committed:
Delete _cutil.cpp

Mrqeoqqt committed:
Delete c_distances_openmp.c

Mrqeoqqt committed:
Delete libdcd.c

Mrqeoqqt committed:
Delete c_distances.c

Mrqeoqqt committed:
Delete libmdaxdr.c

Mrqeoqqt committed:
Delete nsgrid.cpp

Mrqeoqqt committed:
Delete stochasticproxembed.c

Mrqeoqqt committed:
Delete cython_util.c

MingyiXue committed:
early stop loop when searching 'no_charges'; set charge to empty string and add flag to decide whether to add charge attr or not

MingyiXue committed:
rebase

RMeli commented on package/CHANGELOG:
```diff
@@ -13,11 +13,14 @@ The rules for this file:
 ------------------------------------------------------------------------------
-??/??/?? IAlibay, BFedder, inomag, Agorfa, aya9aladdin, shudipto-amin, cbouy, HenokB, umak1106
+??/??/?? IAlibay, BFedder, inomag, Agorfa, aya9aladdin, shudipto-amin, cbouy, 
+umak1106, Mrqeoqqt
```
```suggestion
umak1106, Mrqeoqqt
```
```suggestion
??/??/?? IAlibay, BFedder, inomag, Agorfa, aya9aladdin, shudipto-amin, cbouy, 
         umak1106, Mrqeoqqt
```

RMeli commented on package/CHANGELOG:
```diff
@@ -2831,3 +2834,4 @@ Testsuite
 
 11/12/07 naveen
   * prepared for release outside lab
+  
```
```suggestion

```

RMeli reviewed (changes requested):
A few suggested changes skimming through (you can just accept them directly if you want). 

What is going on with the deleted files?

Mrqeoqqt committed:
try to recover deleted files

Mrqeoqqt committed:
try to recover deleted files

Mrqeoqqt committed:
try to recover deleted files

Mrqeoqqt commented:
@RMeli Hi, Sorry for the mess. I have restored those files. 

I tried to run a local test, then `MDAnalysis` came an error said that "ValueError: numpy.ndarray size changed, may indicate binary incompatibility", so I have to reinstall some packages. But it turns out when I was re-installing `MDAnalysis`, the conda environment compiled C files to my working directory instead of the environment. I didn't find out these C files and they have my laptop info inside until they got pushed to remote repo. 

The rest work is struggling with rebase... I tried to remove those commits but every time I squash those commit locally, they kept accumulating after pushed to the remote repo.

Mrqeoqqt committed:
restore to original

Mrqeoqqt committed:
change mol2parsers and mol2

Mrqeoqqt committed:
modify mol2parser

Mrqeoqqt commented:
Hi,

Sorry that I have another issue. Somehow after I rebase my branch using this command `git rebase upstream/develop`, the files that I touched count changes differently. For example, I only added a few lines, but the 'Files Changed' shows every line is changed. Do you know how to fix this?

Mrqeoqqt committed:
change crlf to lf

Mrqeoqqt committed:
add test_mol2

Mrqeoqqt committed:
remove MOL2 and MOL2Parser

Mrqeoqqt committed:
add MOL2 and MOL2Parser and convert to LF

Mrqeoqqt committed:
convert to LF

RMeli commented:
@Mrqeoqqt don't be sorry, this is quite normal (especially at the beginning; see [xkcd.com/1597](https://xkcd.com/1597/))!

I've seen the new diff in the files changed and it is not very useful since it appears like the whole file has changed. Honestly, I never used `git rebased` (I use `git merge` because it is non-destructive) therefore I have never encountered this problem...

Can you try to see what happens if you merge the `upstream/develop` branch into your branch?
```bash
# https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork
git remote add upstream https://github.com/MDAnalysis/mdanalysis.git
# https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-from-the-command-line
git fetch upstream
git merge upstream/develop
git push
```

I think the problem _might_ stem from [this](https://www.atlassian.com/git/tutorials/merging-vs-rebasing#the-golden-rule-of-rebasing), since you are working on `develop` on your own fork.

Mrqeoqqt commented:
Hi @RMeli  Thank you. I restored those files. It's caused by the difference between CRLF and LF. I work under WSL but used git from windows to push updates, so I didn't notice the default setting is CRLF.

RMeli commented:
@Mrqeoqqt I see, great you spotted it and fixed it! I'll review ASAP now that the diff is reasonable.

Mrqeoqqt commented:
But I am very worried about previous commits, because there's laptop path information in it, like username. Is there a way I can merge those commit so that those version will not be checkout later?

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -97,6 +97,8 @@ class MOL2Parser(TopologyReaderBase):
     .. versionchanged: 2.0.0
        Parse elements from atom types.
+    .. versionchanged:: 2.2.0
+       Optional columns will be assigned with default values when not provided
```
Explicit the chosen default values here as well, so they do show up in the docs.

Mrqeoqqt replied:
@RMeli I added how to deal with optional fields in class docstring, should I added it in versionchanged tag as well?

RMeli replied:
@orbeckst suggestion to add it to the docstring is much better, I think the `versionchanged` does not need this information.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
         resnames = []
         charges = []
-
+        set_charge = True
+        for line in sections['molecule']:
+            if line.strip().lower() == 'no_charges':
+                set_charge = False
+                break
```
From the [MOL2 file standard](https://zhanggroup.org/DockRMSD/mol2.pdf), the 4th line of the `@<TRIPOS>MOLECULE` section specifies the charge. You should be able to condense this in a single line. All supported charge types are uppercase.

Mrqeoqqt replied:
@RMeli Thank you for this one. It's very reasonable and also saves loop complexity.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
-            aid, name, x, y, z, atom_type, resid, resname, charge = a.split()[:9]
+            columns = a.split()
+            if len(columns) >= 9:
+                columns = columns[:9]
+                aid, name, x, y, z, atom_type, resid, resname, charge = columns
```
You can condense this in one line.
```suggestion
                aid, name, x, y, z, atom_type, resid, resname, charge = columns[:9]
```

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+            else:
+                aid, name, x, y, z, atom_type = columns[:6]
+                resid = '1'
+                resname = ''
```
I think @orbeckst said `UNK` was fine here (see https://github.com/MDAnalysis/mdanalysis/issues/3385#issuecomment-1088124131). Any preference @orbeckst ?

orbeckst replied:
See my review. After discussion on Discord the consensus is not *not include resname in the Topology* if it's not in the file.

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -182,7 +209,8 @@ def parse(self, **kwargs):
         attrs.append(Atomnames(np.array(names, dtype=object)))
         attrs.append(Atomtypes(np.array(types, dtype=object)))
-        attrs.append(Charges(np.array(charges, dtype=np.float32)))
+        if set_charge:
```
You will need a test that checks the case when the charge is not set.

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -97,6 +97,8 @@ class MOL2Parser(TopologyReaderBase):
        Bonds attribute is not added if no bonds are present in MOL2 file
     .. versionchanged: 2.0.0
        Parse elements from atom types.
+    .. versionchanged:: 2.2.0
```
add to the documentation at the top a description of what happens when resid, resnm, charge fields are not provided

Mrqeoqqt replied:
Hi @orbeckst I added under `Notes` in the class description. Is that what you mean?

orbeckst replied:
yes

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+                aid, name, x, y, z, atom_type = columns[:6]
+                resid = '1'
+                resname = ''
+                charge = ''
```
Looks nicer if you set `charge = None` and produces more readable code

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+                for i in range(6, len(columns)):
+                    opt_values[i-6] = columns[i]
+                resid, resname, charge = opt_values
+                if len(charge):
```
then use `if charge is None`: this makes clearer what's happening

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+                    opt_values[i-6] = columns[i]
+                resid, resname, charge = opt_values
+                if len(charge):
+                    set_charge = False
```
add a test case where the file does not contain at least one charge and this case is triggered

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
         resnames = []
         charges = []
-
+        set_charge = True
```
rename to `has_charges` to make meaning clearer

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+                                 f" [charge [status_bit]]]]")
+            else:
+                aid, name, x, y, z, atom_type = columns[:6]
+                resid = '1'
```
Can't you use the int 1 here?

Mrqeoqqt replied:
I tried `resid=1` at the very beginning, but it turns out the interface is not compatible with integer, because everything read from line is a strinh.

orbeckst replied:
What is the specific failure if you use the integer?

Mrqeoqqt replied:
Sorry I mixed up. Yes `resid` can be set to 1, I'll fix it. I set to string previously because in LINE 191(original) `resids = np.array(resids, dtype=np.int32)` can convert everything to integer.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+                for i in range(6, len(columns)):
+                    opt_values[i-6] = columns[i]
+                resid, resname, charge = opt_values
+                if len(charge):
+                    set_charge = False
```
I think this is the opposite of what you wanted to achieve? `len(charge)` is `False` when a charge is missing (i.e. when it remains the empty string that you set as default), which is when you would want to set `set_charge` to `False`? In any case, it is untested...

BTW, you can also evaluate the string directly (`if charge:` or `if not charge:`).

From the file format specification it is unclear to me if this is allowed at all, but I'd think it isn't since the format explicitly specify if charges are present or not. If this is the case (probably someone else knows more) you can just skip charges in this whole block.

orbeckst replied:
@RMeli makes a good point. If the file format requires at the beginning of the file to specify if charges are present then we should just rely on this information. @Mrqeoqqt please read the format specification and let us know what should happen in the following cases:

1. MOLECULE section contains NO_CHARGES  as the *charge_type*
   1.  and **no charges** are present in ATOM section
   2. but **charges are present** in ATOM section
2. MOLECULE section does **not** contain NO_CHARGES (i.e., it contains one of the other keywords indicating some sort of charges, see manual)
   1.  and  **charges are present** in ATOM section
   2. but **no charges** are present in ATOM section

Obviously, 1(i) and 2(i) are the expected cases and should just work without any errors or warnings.

Case 2(ii) should definitely raise an exception.

For Case 1(ii) I am not sure: I feel it should  raise an exception (because the content of the file is inconsistent). However, one could also decide to ignore the charges (and issue a warning) or to read the charges nevertheless (I agree with @RMeli that we should not ~~do that~~ read charges in the presence of NO_CHARGES). @Mrqeoqqt what is your opinion here?

Mrqeoqqt replied:
There are multiple charge types (see below). I suppose except from "NO_CHARGE", the other keywords should indicate charges need to be provided. But the document didn't mention how to deal with case1(ii) and case 2(ii) if happens.
> NO_CHARGES, DEL_RE, GASTEIGER, GAST_HUCK, HUCKEL, 
> PULLMAN, GAUSS80_CHARGES, AMPAC_CHARGES, 
> MULLIKEN_CHARGES, DICT_ CHARGES, MMFF94_CHARGES, 
> USER_CHARGES

@orbeckst What about I implement the situations you have described and deal with warnings and errors?

Correctfy: I tested MMFF94 partial charge in `obabel`, and the correct charge model can show up in output.

orbeckst replied:
Case 2(ii) should just raise an exception (`ValueError`).

If there's nothing defined for Case 1(ii) then I would say we also raise a `ValueError` — that's safer than trying to guess intent and interpreting what is essentially a non-conformant file.

This should simplify your code because you can use the analysis of NO_CHARGES to decide how to handle the rest.

orbeckst commented on package/MDAnalysis/coordinates/MOL2.py:
```diff
@@ -139,6 +139,8 @@ class MOL2Reader(base.ReaderBase):
     .. versionchanged:: 2.0.0
        Bonds attribute is not added if no bonds are present in MOL2 file
+    .. versionchanged:: 2.2.0
+       Optional columns will be assigned with default values when not provided
```
This is not really correct for the coordinate reader. It's more accurate to say that you can now read MOL2 files with optional columns omitted. (The *parser* sets default values.) Please adjust.

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +150,34 @@ def parse(self, **kwargs):
+            else:
+                aid, name, x, y, z, atom_type = columns[:6]
+                resid = '1'
+                resname = ''
```
**resname** is not essential for a Topology. Therefore, do not use a default (set it to `None`) and do *not* include it in the Topology.

See https://github.com/MDAnalysis/mdanalysis/issues/3385#issuecomment-1089067563

Mrqeoqqt replied:
@orbeckst Should I not include resnames attrubute if any of the resname is not provided?

orbeckst replied:
To be clear: *if* you have a resname then *do include resname in the topology*.

Mrqeoqqt replied:
Hi @orbeckst, thank you for the clarification. 

Considering the attributes are passed to `Topology` in a batch, would it be better to set `resname=None` for those atoms who didn't provide if other atoms in the file have a resname? Resnames attribute is not included only when all atoms are notprovided?

> To be clear: _if_ you have a resname then _do include resname in the topology_.

orbeckst replied:
I think our comments crossed path. I hadn't seen your question when I posted my last comment.

> Should I not include resnames attrubute if any of the resname is not provided?

The format definition is not clear about the situation where some lines might contain optional data and others don't. It _looks_ as if this was allowed for `@<TRIPOS>ATOM`:
```
atom_id atom_name x y z atom_type [subst_id [subst_name [charge [status_bit]]]]
```
However, that makes little sense for how MDAnalysis sees the data. Therefore, I propose we treat resname in the same way as charge: If even a single ATOM line is missing *resname* then we will not include *any* resname in the topology. I am also inclined to fail with `ValueError` if we find some lines with and others without resname. 

If someone with more experience with working with mol2 files wants to chime in then please do!

orbeckst replied:
I added a https://github.com/MDAnalysis/mdanalysis/issues/3385#issuecomment-1089582380 to the main issue.

Mrqeoqqt replied:
I will go with this one first.

> If even a single ATOM line is missing resname then we will not include any resname in the topology.

orbeckst replied:
> Considering the attributes are passed to Topology in a batch, would it be better to set resname=None for those atoms who didn't provide if other atoms in the file have a resname? Resnames attribute is not included only when all atoms are notprovided?

That seems to be a good idea: collect everything, then check if there are any `None` in the list. If all are `None` then do not include the TopoloAttr. If some are `None`, raise an exception an fail.

orbeckst reviewed (changes requested):
Main comments
- set resid to 1 as a default
- if resname is not given, do *not* include in Topology
- see comments for minor things

orbeckst commented:
> But I am very worried about previous commits, because there's laptop path information in it, like username. Is there a way I can merge those commit so that those version will not be checkout later?

You did a rebase, which altered the history, so you might have already removed the information from GitHub. In any case, we will **squash-merge** your PR so only what you see under file changes will be visible in the main history.

orbeckst commented:
@Mrqeoqqt please have a look at the comments. There's a discussion how to handle NO_CHARGES that needs to be resolved. Ultimately, we want to implement existing standards as closely as possible and to rather fail than guess wrongly.

Mrqeoqqt committed:
modify versionchanged tag; add resnames attribute only when all are provided; add charges attribute when no confliction in keywords

Mrqeoqqt committed:
Merge remote-tracking branch 'upstream/develop' into after_lf

Mrqeoqqt committed:
modify documentation in MOL2Parser

Mrqeoqqt committed:
PEP8

Mrqeoqqt committed:
remove asv.Raise

Mrqeoqqt committed:
modify charge parse

Mrqeoqqt committed:
fix dealing with resnames

Mrqeoqqt committed:
fix dealing with no resnames

Mrqeoqqt committed:
fix PEP8

Mrqeoqqt committed:
fix PEP8

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -87,6 +87,16 @@ class MOL2Parser(TopologyReaderBase):
     Elements are obtained directly from the SYBYL atom types. If some atoms have
     unknown atom types, they will be assigned an empty element record. If all
     atoms have unknown atom types, the elements attribute will not be set.
+    Dealing with optional fields:
```
add a newline above to make it a new paragraph; look at the generated docs (see the ReadtheDocs check and click on the Details link) to see if it looks correct

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -87,6 +87,16 @@ class MOL2Parser(TopologyReaderBase):
+    If "NO_CHARGES" shows up while charge field appears,
+    ValueError will occur;
+    If charge model is specified, while some atoms don't have a charge field,
+    ValueError will occur as well.
```
use reST formatting and sphinx roles
```reST
:exc:`ValueError`
```

orbeckst replied:
To the docs, add a 
```reST
Raises
------
ValueError
     explanation ....
```
section that summarizes what you said here. See https://userguide.mdanalysis.org/stable/contributing_code.html#guidelines-for-writing-docstrings

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,10 +160,41 @@ def parse(self, **kwargs):
+                raise ValueError(f"The mol2 file {self.filename}"
+                                 f" indicates no charges, but charge"
+                                 f" provided in line: {a}.")
+            if (charge is None) and has_charges:
```
could be `elif` (unless you rewrite)

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,10 +160,41 @@ def parse(self, **kwargs):
+                    opt_values[i-6] = columns[i]
+                resid, resname, charge = opt_values
+
+            if charge and (not has_charges):
```
I think you have to say `if charge is not None` because a charge of 0 is legitimate but would count as false. If you need to distinguish on `charge is None` then this can become a nested if
```python
if has_charges:
   if charge is None:
      raise
else:
   if charge is not None:
      raise
```

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -191,12 +235,21 @@ def parse(self, **kwargs):
+            attrs.append(Resnums(resids.copy()))
+        else:
+            raise ValueError(f"Some atoms in the mol2 file {self.filename}"
+                             f" contain subst_name while some not.")
```
grammar "some do not"

orbeckst reviewed (changes requested):
Overall, looks really good, at quick glance I only found a few minor things. EDIT: Please resolve the conflicts, too. (You'll have to do this a couple of times before merge, just because we have quite a few other PRs being merged)

@RMeli what do you think?

Mrqeoqqt committed:
nest logic to deal with charges

Mrqeoqqt committed:
modify docstring in MOL2Parser, add raise error section

Mrqeoqqt committed:
Merge branch 'develop' into develop

Mrqeoqqt committed:
remove appended doc

Mrqeoqqt committed:
add charge is not None

Mrqeoqqt committed:
before change if charge and has_charge logic

Mrqeoqqt committed:
add parathesis

RMeli commented on package/CHANGELOG:
```diff
@@ -13,11 +13,13 @@ The rules for this file:
 
 ------------------------------------------------------------------------------
-??/??/?? IAlibay, BFedder, inomag, Agorfa, aya9aladdin, shudipto-amin, cbouy, HenokB, umak1106, tamandeeps
+??/??/?? IAlibay, BFedder, inomag, Agorfa, aya9aladdin, shudipto-amin, cbouy, HenokB, umak1106, tamandeeps, Mrqeoqqt
```
Not your fault @Mrqeoqqt, but I think this should go in a new line? @orbeckst what is the policy about fixing the CHANGELOG?
```suggestion
??/??/?? IAlibay, BFedder, inomag, Agorfa, aya9aladdin, shudipto-amin, cbouy,
         HenokB, umak1106, tamandeeps, Mrqeoqqt
```

orbeckst replied:
Fixing the line length is totally fine.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -87,6 +87,36 @@ class MOL2Parser(TopologyReaderBase):
+    ------
+    ValueError
+      If some atoms have the optional field ``resname`` (aka ``subst_name``)
+      while some do not, this error would be raised.
```
```suggestion
      while some do not.
```
I think the last part is redundant. The way I read (and try to write these docstrings is): `Raises ValueError if some atoms have [...]`

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -87,6 +87,36 @@ class MOL2Parser(TopologyReaderBase):
+    ValueError
+      If "NO_CHARGES" shows up in "@<TRIPOS>MOLECULE" section while
+      some atoms have the optional field ``charge`` (aka ``subst_name``),
+      this error would be raised.
```
`charge` and `subst_name` are two different fields:
```suggestion
      some atoms have the optional field ``charge``.
```

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -87,6 +87,36 @@ class MOL2Parser(TopologyReaderBase):
+    ValueError
+      If "NO_CHARGES" does not up in "@<TRIPOS>MOLECULE" section while
+      some atoms do not have the optional field ``charge`` (aka ``subst_name``),
+      this error would be raised.
```
```suggestion
      If a charge model is specified in "@<TRIPOS>MOLECULE" section while
      some atoms do not have the optional field ``charge``.
```

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +180,51 @@ def parse(self, **kwargs):
         charges = []
+        has_charges = True
+        if sections['molecule'][3].strip() == 'NO_CHARGES':
+            has_charges = False
```
This can be condensed in one line.

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +180,51 @@ def parse(self, **kwargs):
+                                 f" [charge [status_bit]]]]")
+            else:
+                aid, name, x, y, z, atom_type = columns[:6]
+                resid = 1
+                resname = None
+                charge = None
+                opt_values = [resid, resname, charge]
+                for i in range(6, len(columns)):
+                    opt_values[i-6] = columns[i]
+                resid, resname, charge = opt_values
```
I'm wondering if this can be made a bit more compact and readable, but I am not sure. Maybe something like
```suggestion
                id_name_charge = [1, None, None]
                for i, col in enumerate(range(6, len(columns))):
                    opt_values[i] = columns[col]
                resid, resname, charge = id_name_charge
```
What do you think @orbeckst @Mrqeoqqt?

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +180,51 @@ def parse(self, **kwargs):
+            #                          f" no charge provided in line: {a}")
+            if charge is not None and (not has_charges):
+                raise ValueError(f"The mol2 file {self.filename}"
+                                 f" indicates no charges, but charge"
```
```suggestion
                                 f" indicates no charges, but charge {charge}"
```
Maybe worth printing out the value of the charge as well?

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -191,12 +266,21 @@ def parse(self, **kwargs):
+            attrs.append(Resids(resids))
+            attrs.append(Resnums(resids.copy()))
+            attrs.append(Resnames(resnames))
+        elif not (np.any(resnames)):
```
```suggestion
        elif not np.any(resnames):
```

RMeli reviewed (commented):
A few more suggestions, mostly on docs and output strings. I think we are nearly there! =)

Mrqeoqqt committed:
fic docstring

Mrqeoqqt committed:
remove redundancy in docstring

Mrqeoqqt committed:
remove redundancy in docstring

Mrqeoqqt committed:
fix PEP8

Mrqeoqqt committed:
change logic of 'has_charge'

Mrqeoqqt committed:
lowercase resid

Mrqeoqqt commented:
@orbeckst @RMeli Thank you for the comments on docstring. I tested the sphinx doc, and the following is the output of my changes. Is this OK?

![image](https://user-images.githubusercontent.com/19585264/162258238-5da66a1d-b3d5-4490-a034-917754a9268a.png)


@orbeckst I changed the if-else logic suggested by you. The logic is much clear than what I have done.

> I think you have to say `if charge is not None` because a charge of 0 is legitimate but would count as false. If you need to distinguish on `charge is None` then this can become a nested if
> 
> ```python
> if has_charges:
>    if charge is None:
>       raise
> else:
>    if charge is not None:
>       raise
> ```

orbeckst commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -88,15 +88,44 @@ class MOL2Parser(TopologyReaderBase):
     atoms have unknown atom types, the elements attribute will not be set.
 
+    Dealing with optional fields:
+    (1) ``resid`` will set to 1 when not provided.
+    (2) If no atoms have ``resname`` field, resnames attribute will not be set;
```
If you want this to be formatted as an enumerated list then you need to use the proper reST formatting.

Mrqeoqqt replied:
@orbeckst Thank you for the comment. I intended to make an enumerate list, but not very familiar with ReST. When I type the following (according to ReST doc), it didn't output the desired indentation before the 2nd and 3nd (see image).
```
"""
    Dealing with optional fields:
    
    1. ``resid`` will set to 1 when not provided.
        
    2. If no atoms have ``resname`` field, resnames attribute will not be set;
    If some atoms have ``resname`` while some do not,
    :exc:`ValueError` will occur.
    
    3. If "NO_CHARGES" shows up in "@<TRIPOS>MOLECULE" section
    and no atoms have the ``charge`` field, charges attribute will not be set;
    If "NO_CHARGES" shows up while ``charge`` field appears,
    :exc:`ValueError` will occur;
    If a charge model is specified while some atoms don't have ``charge``,
    :exc:`ValueError` will occur as well.
"""
```
![image](https://user-images.githubusercontent.com/19585264/162272330-f28049e1-c36b-46ce-9cf0-86b7db3f572a.png)


However, if without space lines between different points, `build_sphinx` will raise an error. Sorry that I don't know how to make the enumerate list work.
```Warning, treated as error:
MOL2Parser.py:docstring of MDAnalysis.topology.MOL2Parser.MOL2Parser:25:Enumerated list ends without a blank line; unexpected unindent.
```

Mrqeoqqt committed:
fix PEP8

orbeckst reviewed (approved):
Great work, you did a thorough job making the parser more robust with sensible error messages and comprehensive tests. The docs look good, too. 

I have a minor optional comment but I am happy with the PR as it stands.

@RMeli what's your opinion? I'll wait for you before I squash-merge.

Mrqeoqqt committed:
fix enumerated list

orbeckst commented:
I think you need to indent subsequent lines as in 
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks

> On Apr 7, 2022, at 11:54 AM, Mingyi Xue ***@***.***> wrote:
> 
> 
> @Mrqeoqqt commented on this pull request.
> 
> In package/MDAnalysis/topology/MOL2Parser.py <https://github.com/MDAnalysis/mdanalysis/pull/3598#discussion_r845469967>:
> 
> > +    (1) ``resid`` will set to 1 when not provided.
> +    (2) If no atoms have ``resname`` field, resnames attribute will not be set;
> @orbeckst <https://github.com/orbeckst> Thank you for the comment. I intended to make an enumerate list, but not very familiar with ReST. When I type the following (according to ReST doc), it didn't output the desired indentation before the 2nd and 3nd (see image).
> 
> """
>     Dealing with optional fields:
>     
>     1. ``resid`` will set to 1 when not provided.
>         
>     2. If no atoms have ``resname`` field, resnames attribute will not be set;
>     If some atoms have ``resname`` while some do not,
>     :exc:`ValueError` will occur.
>     
>     3. If "NO_CHARGES" shows up in "@<TRIPOS>MOLECULE" section
>     and no atoms have the ``charge`` field, charges attribute will not be set;
>     If "NO_CHARGES" shows up while ``charge`` field appears,
>     :exc:`ValueError` will occur;
>     If a charge model is specified while some atoms don't have ``charge``,
>     :exc:`ValueError` will occur as well.
> """
>  <https://user-images.githubusercontent.com/19585264/162272330-f28049e1-c36b-46ce-9cf0-86b7db3f572a.png>
> However, if without space lines between different points, build_sphinx will raise an error. Sorry that I don't know how to make the enumerate list work.
> 
> MOL2Parser.py:docstring of MDAnalysis.topology.MOL2Parser.MOL2Parser:25:Enumerated list ends without a blank line; unexpected unindent.
> —
> Reply to this email directly, view it on GitHub <https://github.com/MDAnalysis/mdanalysis/pull/3598#discussion_r845469967>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAB2DHF33AM47CX2KCLPLZDVD4VPBANCNFSM5SMTHCAA>.
> You are receiving this because you were mentioned.
> 

--
Oliver Beckstein (he/his/him)
***@***.***

Mrqeoqqt commented:
Now it works :)
![image](https://user-images.githubusercontent.com/19585264/162327785-81652701-2f08-4475-8919-ce6568d9de26.png)

RMeli commented on package/MDAnalysis/topology/MOL2Parser.py:
```diff
@@ -148,9 +180,37 @@ def parse(self, **kwargs):
         resnames = []
         charges = []
-
+        has_charges = False if sections['molecule'][3].strip() == 'NO_CHARGES'\
+            else True
```
What I was thinking here was
```suggestion
        has_charges = sections['molecule'][3].strip() != 'NO_CHARGES'
```

When you have `True if CONDITION else False` you can work directly with `CONDITION` since it already returns a `bool`.

RMeli reviewed (approved):
All my comments have been addressed, thank you @Mrqeoqqt ! The last comment is not critical.

@orbeckst I'll let you have the last look and squash-merge.

Mrqeoqqt committed:
no_charge in one line

orbeckst reviewed (approved):
nice work, all looking good

I'll merge right away.

orbeckst commented:
Congratulations @Mrqeoqqt !! Thank you for your first contribution.

Thank you @RMeli for the fantastic code reviews.

RMeli commented:
Congratulations @Mrqeoqqt, great work!
